### PR TITLE
8321137: Reconsider ICStub alignment

### PR DESCRIPTION
**Pull request by shipilev**
This continues from #16911. It initially started as performance optimization to compact `ICStubs`, but I think the safety arguments for fitting the `ICStub` per instruction cache line prevails. See bug and previous PR for more gory details. The footprint improvements on some architectures come as side-effect of untying the `ICStub` size from `CodeEntryAlignment` to (sometimes lower) cache line size.

Note that the size of `ICStub` is important, because `ICBuffer` is small (10K by default), and its depletion causes the `ICBufferFull` safepoint. I would make a (separate) argument to bump the default `ICBuffer` size a bit to make it less important.

Current patch affects `ICStub` size in different ways on different platforms, since current size is effectively 2x`CodeEntryAlignment` and new size is cache line size:
 - ARM32: 32 -> 64 bytes :(
 - AArch64: 128 -> 64 bytes :)
 - x86_64: 64 -> 64 bytes :|
 - x86_32: 32 -> 64 bytes :(
 - PPC64: 512 -> 128 bytes :))
 - RISC-V: 128 -> 64 bytes :)
 - S390X: 128 -> 256 bytes :(
 - Zero: <not applicable, Zero does not use IC stubs>

Additional testing:
 - [x] Linux x86_64 server fastdebug, `tier{1,2,3,4}`
 - [x] Linux AArch64 server fastdebug, `tier{1,2,3,4}`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321137](https://bugs.openjdk.org/browse/JDK-8321137): Reconsider ICStub alignment (**Enhancement** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [b0875b50](https://git.openjdk.org/jdk/pull/17277/files/b0875b50dc88934c48b48e410d43730b86960102)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [b0875b50](https://git.openjdk.org/jdk/pull/17277/files/b0875b50dc88934c48b48e410d43730b86960102)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [b0875b50](https://git.openjdk.org/jdk/pull/17277/files/b0875b50dc88934c48b48e410d43730b86960102)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17277/head:pull/17277` \
`$ git checkout pull/17277`

Update a local copy of the PR: \
`$ git checkout pull/17277` \
`$ git pull https://git.openjdk.org/jdk.git pull/17277/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17277`

View PR using the GUI difftool: \
`$ git pr show -t 17277`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17277.diff">https://git.openjdk.org/jdk/pull/17277.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17277#issuecomment-1878535395)

### CODE REVIEW

**Review by dean-long — APPROVED** [review]
Nice job minimizing the changes.

File: src/hotspot/share/code/icBuffer.cpp [hunk 1]
```diff
@@ -219,9 +219,18 @@ bool InlineCacheBuffer::create_transition_stub(CompiledIC *ic, void* cached_valu
...
   }
 
+#ifdef ASSERT
+  {
+    ICStub* rev_stub = ICStub::from_destination_address(ic_stub->code_begin());
+    assert(ic_stub == rev_stub,
+           "ICStub mapping is reversible: stub=" PTR_FORMAT ", code=" PTR_FORMAT ", rev_stub=" PTR_FORMAT,
+           p2i(ic_stub), p2i(ic_stub->code_begin()), p2i(rev_stub));
+  }
+#endif
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/code/icBuffer.cpp:224]: Problem is, `new_ic_stub` can return null on out of memory, so we would need to check that. But I think `new_ic_stub` does not carry its weight, so I just inlined it in new commit, which looks like a good middle ground?

[the remark this answers is not in the document]
Comment (dean-long) [reply src/hotspot/share/code/icBuffer.cpp:224]: Sure, that's fine.  I still feel like from_destination_address sanity checks should not be in the caller, but should be done at ICStub creation time or when code_begin() is called, but it's not a big deal I guess if ICStub can go away soon.

**Review by dean-long — APPROVED** [review]

**Comment by shipilev** [comment]
Thanks @dean-long! Any other takers?

Notably, I would like opinions of platform maintainers: @theRealAph, @TheRealMDoerr, @RealFYang, @bulasevich.

**Comment by TheRealMDoerr** [comment]
It's probably good, but are you aware of the plans to remove the ICStubs completely? @fisk may have an opinion on it.

**Comment by shipilev** [comment]
> It's probably good, but are you aware of the plans to remove the ICStubs completely? @fisk may have an opinion on it.

Yes, I did x86_32 version for Erik. I think we want something we can support in JDK update releases. I think the ICStub removal would not be backportable at all.

**Comment by TheRealMDoerr** [comment]
Ok. I think it's good for backports. I'll put it in our test queue.

**Review by fisk — APPROVED** [review]
Seems like a reasonable band-aid to backport. Looks good.

**Review by TheRealMDoerr — APPROVED** [review]
Test results are good. Thumbs up from my side!

File: src/hotspot/share/code/icBuffer.hpp [hunk 2]
```diff
@@ -62,16 +62,27 @@ class ICStub: public Stub {
...
   int     size() const                           { return _size; }
 
-  // ICStub_from_destination_address looks up Stub* address from code entry address,
-  // which unfortunately means the stub head should be at the same alignment as the code.
-  static  int alignment()                        { return CodeEntryAlignment; }
+  // For extra correctness/safety, we want to make sure that each ICStub is in
+  // a separate instruction cache line. This would allow for piggybacking on instruction
+  // cache coherency on some architectures to order the updates to ICStub and setting
+  // the destination to the ICStub. Note that cache lines size might be larger than
+  // CodeEntryAlignment that is a normal alignment for CodeBlobs.
```
Comment (theRealAph) [inline src/hotspot/share/code/icBuffer.hpp:69]: This is a paragraph that raises more questions than it answers. The relationship between ordering and instruction cache line coherency is pretty tenuous even on x86 these days, isn't it?  I'd not say "For extra correctness/safety," but "To be cautious,".

Comment (shipilev) [reply src/hotspot/share/code/icBuffer.hpp]: Sure. How does the new comment look?

**Review by RealFYang — APPROVED** [review]
LGTM. Also performed tier1-3 test with fastdebug build on linux-riscv64 platform.

**Review by theRealAph — APPROVED** [review]

**Comment by shipilev** [comment]
OK, thanks! I am going to merge it, and look for follow-ups, if any.

/integrate

**Pull request closed without merging**